### PR TITLE
issue-8783: handling pristine default security groups

### LINKED ISSUE #8783: the unused filter for aws.security-group needs to ignore default IpPermissions
### Describe the bug

A [default security group](https://docs.aws.amazon.com/vpc/latest/userguide/default-security-group.html) has the following IpPermission when the group is created:

```
                {
                    "IpProtocol": "-1",
                    "IpRanges": [],
                    "Ipv6Ranges": [],
                    "PrefixListIds": [],
                    "UserIdGroupPairs": [
                        {
                            "GroupId": "sg-<ID number>",
                            "UserId": <account ID>"
                        }
                    ]
                }
```

This is to allow same-account traffic to members of that security group. This poses a problem for us when we try to use the `type: unused` filter for default security groups. When the filter [scans for used groups](https://github.com/cloud-custodian/cloud-custodian/blob/da894ebc7b4a1b829845860378b58e714c6fb53f/c7n/resources/vpc.py#L885-L887), it [uses the get_sg_refs() method as a scanner](https://github.com/cloud-custodian/cloud-custodian/blob/da894ebc7b4a1b829845860378b58e714c6fb53f/c7n/resources/vpc.py#L877), which [iterates through these rules](https://github.com/cloud-custodian/cloud-custodian/blob/da894ebc7b4a1b829845860378b58e714c6fb53f/c7n/resources/vpc.py#L933-L934). If any of these rules [has a UserGroupIdPair at all](https://github.com/cloud-custodian/cloud-custodian/blob/da894ebc7b4a1b829845860378b58e714c6fb53f/c7n/resources/vpc.py#L936C49-L936C49), then the group [will be considered as used](https://github.com/cloud-custodian/cloud-custodian/blob/da894ebc7b4a1b829845860378b58e714c6fb53f/c7n/resources/vpc.py#L937-L938). The filter needs to take such rules into account because it is not necessarily true that a group with these rules (which AWS generates automatically) is used.

### What did you expect to happen?

I expect a policy similar to the following:

```yaml
policies:
  - name: enforce-empty-unused-default-sgs-all-vpcs
    description: ensure that the unused default security groups of all VPCs are empty
    resource: aws.security-group
    filters:
      - type: unused
      - type: value
        key: GroupName
        op: equal
        value: "default"
```

to match all of the security groups named `default` that are unused, regardless of the fact that they have rules automatically generated by AWS.

### Cloud Provider

Amazon Web Services (AWS)

### Cloud Custodian version and dependency information

```shell
c7n==0.9.26
```

### Policy

```shell
policies:
  - name: enforce-empty-unused-default-sgs-all-vpcs
    description: ensure that the unused default security groups of all VPCs are empty
    resource: aws.security-group
    filters:
      - type: unused
      - type: value
        key: GroupName
        op: equal
        value: "default"
```

### Relevant log/traceback output

_No response_

### Extra information or context

_No response_

### PR DESCRIPTION
**Pull request by cvirtucio**
# Description

This resolves #8783.

### CODE REVIEW

**Comment by kapilt** [comment]
`make lint` which you can run locally via the same, is reporting a few errors re ci.. 

```
Run make lint
ruff c7n tests tools
Error: c7n/resources/vpc.py:1019:101: E501 Line too long (135 > 100 characters)
make: *** [Makefile:67: lint] Error 1
Error: c7n/resources/vpc.py:1053:101: E501 Line too long (102 > 100 characters)
Error: c7n/resources/vpc.py:1057:101: E501 Line too long (101 > 100 characters)
Error: c7n/resources/vpc.py:1062:101: E501 Line too long (108 > 100 characters)
Error: c7n/resources/vpc.py:1066:101: E501 Line too long (107 > 100 characters)
Error: Process completed with exit code 2.
```

**Comment by cvirtucio** [comment]
> `make lint` which you can run locally via the same, is reporting a few errors re ci..
> 
> ```
> Run make lint
> ruff c7n tests tools
> Error: c7n/resources/vpc.py:1019:101: E501 Line too long (135 > 100 characters)
> make: *** [Makefile:67: lint] Error 1
> Error: c7n/resources/vpc.py:1053:101: E501 Line too long (102 > 100 characters)
> Error: c7n/resources/vpc.py:1057:101: E501 Line too long (101 > 100 characters)
> Error: c7n/resources/vpc.py:1062:101: E501 Line too long (108 > 100 characters)
> Error: c7n/resources/vpc.py:1066:101: E501 Line too long (107 > 100 characters)
> Error: Process completed with exit code 2.
> ```

fixed: 90c5fb1a4e39d6657d47042f0f06b4aabc489156

```txt
(.venv) christopher.virtucio@KGYF7CQGCT cloud-custodian % make lint
ruff c7n tests tools
black --check tools/c7n_left tools/c7n_mailer tools/c7n_oci tools/c7n_kube
All done! ✨ 🍰 ✨
127 files would be left unchanged.
type -P terraform && terraform fmt -check -recursive .
/opt/homebrew/bin/terraform
(.venv) christopher.virtucio@KGYF7CQGCT cloud-custodian %
```

File: c7n/resources/vpc.py [hunk 1]
```diff
@@ -1012,6 +1015,92 @@ def get_batch_sgs(self):
         resources = self.manager.get_resource_manager('aws.batch-compute').resources(augment=False)
         return set(expr.search(resources) or [])
 
+    def is_pristine_default_sg(self, account_id, sg):
```
[the remark this answers is not in the document]
Comment (kapilt) [reply c7n/resources/vpc.py:1018]: I don't really think we should consider references from a group to itself as usage, which means we can do a one liner for the fix.

**Comment by kapilt** [comment]
I put up #8821 which should cover the more general case of not considering a group that references itself as usage.

**Pull request closed without merging**